lib/int64.js

```diff
@@ -0,0 +1,335 @@
+        res.value[0] = val0;
+        res.value[1] = (val1 << 1) | ((val0 >> 31) & 1);
+      } else if (value.startsWith('0b')) {
+        value = value.slice(2);
+        res.value[0] = Number('0b' + value.slice(value.length - 32));
+        res.value[1] = Number('0b' + value.slice(0, value.length - 32));
+      } else {
+        const sign = value.startsWith('-');
+        if (sign) value = value.slice(1);
+        const intRadix = new Int64(10);
+        for (let i = 0; i < value.length; i++) {
+          const intDigit = new Int64(value[i]);
+          res = Int64.mult(res, intRadix).add(intDigit);
+        }
+        if (sign) res.value = Int64._conversion(res.value);
+      }
+      this.value[0] = res.value[0];
+      this.value[1] = res.value[1];
+    }
+  }
+
+  toInt32() {
+    return this.value[0];
+  }
+
+  static _conversion(value) {
+    const result = new Uint32Array(value);
+    let pos = 0;
+    if (result[0] === 0) {
+      while (pos < 32) {
+        if ((result[1] >>> pos) & 1) break;
+        pos++;
+      }
+      const mask = (1 << (pos + 1)) - 1;
+      result[1] = (~(result[1] | mask)) | (result[1] & mask);
+    } else {
+      while (pos < 32) {
+        if ((result[0] >>> pos) & 1) break;
+        pos++;
+      }
+      const mask = (1 << (pos + 1)) - 1;
+      result[0] = (~(result[0] | mask)) | (result[0] & mask);
+      result[1] = ~result[1];
+    }
+    return result;
+  }
+
+  add(b) {
+    const tmp = this.value[0] + b.value[0];
+    this.value[0] = tmp;
+    this.value[1] += b.value[1] + (tmp > UINT32_MAX ? 1 : 0);
+    return this;
+  }
+
+  static add(a, b) {
+    return new Int64(a).add(b);
+  }
+
+  sub(b) {
+    let value = new Uint32Array(b.value);
+    if (value[0] !== 0 || value[1] !== 0) value = Int64._conversion(value);
+    const tmp = this.value[0] + value[0];
+    this.value[0] = tmp;
+    this.value[1] += value[1] + (tmp > UINT32_MAX ? 1 : 0);
+    return this;
+  }
+
+  static sub(a, b) {
+    return new Int64(a).sub(b);
+  }
+
+  static cmp(a, b) {
+    const aSign = a.value[1] >>> 31;
+    const bSign = b.value[1] >>> 31;
+    if (aSign !== bSign) return aSign ? -1 : 1;
+
+    if (a.value[1] > b.value[1]) return 1;
+    if (a.value[1] < b.value[1]) return -1;
+    if (a.value[0] === b.value[0]) return 0;
+    return a.value[0] > b.value[0] ? 1 : -1;
+  }
+
+  static _division(n, d) {
+    const zero = new Int64();
+    const one = new Int64(1);
+    if (Int64.cmp(d, zero) === 0) {
+      throw new RangeError('Int64: division by zero');
+    }
+    if (Int64.cmp(n, zero) === 0) return [zero, 0];
+    const cmp = Int64.cmp(d, n);
+    if (cmp === 0) return [one, zero];
+
+    const nSign = n.value[1] >>> 31;
+    let value = new Uint32Array(n.value);
+    if (nSign) value = Int64._conversion(value);
+    const dSign = d.value[1] >>> 31;
+    const dPositive = new Int64(d);
+    if (dSign) dPositive.value = Int64._conversion(dPositive.value);
```

Comment:
IIUC this is converting negative num to positive number representation in order to divide them as unsigned. If that’s the case I’d want to actually implement signed-division instead of converting them here and at the end.